### PR TITLE
Metadata

### PR DESCRIPTION
**Pull request by KurtMi**
Add the meta script to organize metadata of scripts and add some metadata to some scripts.

### CODE REVIEW

File: scripts/configure-debian [hunk 1]
```diff
@@ -1,4 +1,10 @@
 #!/bin/sh
+#
+# @author Markus Raab <elektra@markus-raab.org>
+# @desc Configure a debain build
```
Comment (markus2330) [inline scripts/configure-debian]: debian

File: scripts/configure-debian [hunk 2]
```diff
@@ -1,4 +1,10 @@
...
+# @date 02.09.2014
+# @tags configure
+# @system debain
```
Comment (markus2330) [inline scripts/configure-debian]: copy&paste

File: scripts/configure-debian [hunk 3]
```diff
@@ -1,4 +1,10 @@
 #!/bin/sh
+#
+# @author Markus Raab <elektra@markus-raab.org>
+# @desc Configure a debain build
+# @date 02.09.2014
+# @tags configure
```
Comment (markus2330) [inline scripts/configure-debian]: What about auto-tags if its a part of the applications name?

File: scripts/copy-template [hunk 4]
```diff
@@ -1,4 +1,10 @@
 #!/bin/sh
+#
+# @author Markus Raab <elektra@markus-raab.org>
+# @desc Creates a plugin out of the template plugin
+# @date 05.05.2012
+# @tags template
```
Comment (markus2330) [inline scripts/copy-template]: `@tags generator creator`

File: scripts/meta [hunk 5]
```diff
@@ -0,0 +1,162 @@
+#!/usr/bin/env python
+# vim: noexpandtab:tabstop=4:shiftwidth=4:fileencoding=UTF-8
+#
+# @author Sebastian Bachmann <sebastian.bachmann@tuwien.ac.at>
```
Comment (markus2330) [inline scripts/meta]: I am not sure if its ok to use his e-mail address here. Is he on github?

File: scripts/meta [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+#
+# Comments in the Beginning of a script define some things and are parsed by this meta script
+# The aim of this script is to find the right script in no time!
+#
+# The Metatag System of Epydoc is used (http://epydoc.sourceforge.net/manual-fields.html#module-metadata-variables)
+# and extendet with the following special tags:
+# 
+# MetaTag					Meaning
+# ----------------------------------------------------------
+# @date						The Date when the script was created, use DD.MM.YYYY as format
```
Comment (markus2330) [inline scripts/meta]: avoid usage of tabs (wrongly displayed on github).

Can you create a README.md for how to tag scripts?

File: scripts/meta [hunk 7]
```diff
@@ -0,0 +1,162 @@
+#!/usr/bin/env python
+# vim: noexpandtab:tabstop=4:shiftwidth=4:fileencoding=UTF-8
+#
+# @author Sebastian Bachmann <sebastian.bachmann@tuwien.ac.at>
+# @desc This script is used to organize scripts
+# @date 10.05.2016
+# @tags
+# @system all
+#
+# Comments in the Beginning of a script define some things and are parsed by this meta script
```
Comment (markus2330) [inline scripts/meta]: beginning, "some things"?

File: scripts/meta [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+# The aim of this script is to find the right script in no time!
+#
+# The Metatag System of Epydoc is used (http://epydoc.sourceforge.net/manual-fields.html#module-metadata-variables)
```
Comment (markus2330) [inline scripts/meta]: shouldn't we prefer doxygen?

**Comment by markus2330** [comment]
Thank you! The script does not seem to be installed? Does it work when installed with `kdb meta`?

File: scripts/meta [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+# @depends					Other script names that depend on (the output) of this script
+
+metatags = ["date", "desc", "author", "tags", "system", "depends"]
+
+# 
+# Because there are not only python scripts, we use the @ declaration, inside the comment.
+# Beware, that these metatags should be applied at the beginning of the file (in the first 10 rows)
+#
+# If you create own Tags, please document them here! 
+tags = {'configure': "This script is used for the build configuration",
```
Comment (markus2330) [inline scripts/meta]: again tab usage does not align (but should be used here), can you switch to our defaults as given in .clang-format

File: scripts/meta [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+		head = map(lambda x: x.replace("\n", "").replace("\r", ""), list(islice(myfile, n)))
+	return head
+
+
+def get_scripts():
+	scripts = defaultdict(dict)
+	scripts_path = os.path.dirname(os.path.realpath(__file__))
+
+	# FIXME we believe that every script uses a hash as comment marker...
+	tag_match = re.compile("^#[ ]*@(%s) (.*)$" % "|".join(metatags))
```
Comment (markus2330) [inline scripts/meta]: try to detect file type in first line?

Comment (markus2330) [inline scripts/meta]: Alternatively, the python-magic libray could be interesting.

**Comment by KurtMi** [comment]
Btw there are issues after the install.

**Comment by KurtMi** [comment]
Now it should be fine. Some notes:

* python-magic libray for file type detection was avoided, due to dependencies to another lib
* tabs should be fine now, but I am not sure, this tab and space topic combined with the python indentation confuses me.
* doxygen was not used, because for now doxygen parses only code and markdown files, using doxygen for script files would be to much config overhead, in my opinion (but I made a Readme.md yay).
* auto tagging seems to intransparent

File: scripts/README.md [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+| \@depends | Other script names that depend on (the output) of this script        |
+
+Do not mind the '\' at the beginning it is a doxygen escaping.
+
+Because there are not only python scripts, we use the @ declaration, inside the comment.
+Beware, that these metatags should be applied at the beginning of the file (in the first 10 rows)
+
+## Tags ##
+
+If you choose to add a tag to the `\@tags` then do not forget to add it in the tags map of the `meta` script.
```
Comment (markus2330) [inline scripts/README.md]: is escaping of @ necessary here?

File: scripts/README.md [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+Just set the Tags as described above as comments at the start (within 10 lines) of your script.
+
+
+Example from `meta`:
+
+```
+<Start of File>
+#!/usr/bin/env python
+#
+# @author Sebastian Bachmann support by <kurt.micheli@libelektra.org>
```
Comment (markus2330) [inline scripts/README.md]: We should somehow standardize multiauthors.

File: scripts/README.md [hunk 13]
```diff
@@ -0,0 +1,56 @@
+# Kdb Meta #
```
Comment (markus2330) [inline scripts/README.md]: Can you make man page out of it? (see getenv)

File: scripts/README.md [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+
+Meta tags as comments in the beginning of a script are parsed by the meta script.
```
Comment (markus2330) [inline scripts/README.md]: Important here would be how to search using meta

File: scripts/README.md [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+The aim of `kdb meta` is to find the right script in no time!
+
+The Metatag System of Epydoc is used (http://epydoc.sourceforge.net/manual-fields.html#module-metadata-variables)
+and extendet with the following special tags:
```
Comment (markus2330) [inline scripts/README.md]: extended

**Comment by markus2330** [comment]
I do not understand your reasoning behind doxygen. Maybe its an misunderstanding: I mean use the keywords of doxygen, not the tool.

Nearly all of the extended keywords are part of doxygen.

**Comment by KurtMi** [comment]
Ah ok, so @desc should be @brief?

**Comment by markus2330** [comment]
Exactly. I think documenting every tag to be used is useful anyway. I think its better to have a small set of tags which is used throughly.

**Comment by KurtMi** [comment]
Ok, in my opinion the @system and @depends tag are not so usefull. system tags can be simulated with a @tags tag like:
```
@tags config, osx, build
```
And the case of dependencies is really rare in this small set of scripts.

**Comment by markus2330** [comment]
I fully agree.

**Comment by KurtMi** [comment]
And we should rename the whole thing to scripts or so. Because `kdb meta` is colliding with the meta data of a Key and its command line syntax.

**Comment by markus2330** [comment]
Yes, you are right.

Maybe `find-tools`?

**Comment by KurtMi** [comment]
`find-tools` sounds nice

**Comment by markus2330** [comment]
@reox can you take a look please?

**Comment by reox** [comment]
Hi!
maybe remove the glob:

$ pyflakes find-tools
find-tools:21: 'glob' imported but unused

pep8 gives me a lot of warnings, could be changed for better code-style (also tabs vs spaces...)

otherwise seems good!

**Comment by markus2330** [comment]
Thanks for the review!

@KurtMi can you remove the glob and fix the non-tab warning of pep8?

**Comment by reox** [comment]
@markus2330 pep8 states that spaces should be used, so a lot of warnings is generated. but you can ignore them i think... but there are other warnings like whitespace before operator and so on. I did not posted the list of warnings here because it is quite long (and there is probably no urge to fix them)

**Comment by markus2330** [comment]
Yes, I know that pep8 states spaces should be used. But Elektra uses tabs, see doc/CODING.md. That is the reason I asked for only fixing non-tab related warnings (if anything at all, @KurtMi usually is very careful in formatting and knows the Elektra's coding guidelines).

Would be good if I can merge this soon, I would like to build debian packages which would be difficult without merging. (Because it needs also changes in the debian branch).

**Pull request merged**